src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/TestUtil.java

```diff
@@ -237,4 +251,22 @@ public static int adjustFlushCountForOutputStreamClose(int flushCount) {
         return result;
     }
 
+    /**
+     * Runs a given task with limited stack size and time duration, then
+     * returns the result. See {@link AbstractModuleTestSupport#verifyWithLimitedResources}
+     * for an example of how to use this method when task does not return a result, i.e.
+     * the given method's return type is {@code void}.
+     *
+     * @param callable the task to execute
+     * @param <V> return type of task - {@code Void} if task does not return result
+     * @return result
+     * @throws Exception if getting result fails
+     */
+    public static <V> V getResultWithLimitedResources(Callable<V> callable) throws Exception {
+        final FutureTask<V> futureTask = new FutureTask<>(callable);
+        final Thread thread = new Thread(null, futureTask,
+                "LimitedStackSizeThread", MINIMAL_STACK_SIZE);
+        thread.start();
+        return futureTask.get(10, TimeUnit.SECONDS);
+    }
```

Comment:
Ok, now we can reuse this method anywhere that we need to run a test with limited stack size and time duration. See https://github.com/checkstyle/checkstyle/pull/10640/files#r690383859

Reply:
So we can do something similar to https://github.com/checkstyle/checkstyle/issues/9253#issuecomment-774761543 and not see a negative impact?

Reply:
> I believe executing junits in this fashion is not advisable. It appears to conflict with static variables and passing expected exceptions along.

@rnveach , is it a negative impact ?

Reply:
@romani I have resolved my own concerns. I updated my branch, https://github.com/rnveach/checkstyle/commits/issue_9253, with this code, and I was able to successfully run all tests with this. My concerns came before when tests were failing with this type of hack in place.

From what I can see, the only nuance to this is how we will have to handle expected exceptions. See https://github.com/rnveach/checkstyle/commit/565182e619a6e87b1ff360bfe020b59a1c06c53c#diff-736846d76eeb3f1bbdd77ef87a82f2c54df2be74058b302f2d6b73e71db3de51R129-R150

If this works, I think we should use it everywhere as we will never be sure where the issue will occur.

Reply:
Here is my attempt to use this approach:

https://github.com/checkstyle/checkstyle/commit/d78f311a2ade3fac24225cad509ee0e2f5527422

Reply:
> Here is my attempt to use this approach

@pbludov I like your attempt, and think that we should proceed with it. This is why:

1. Using annotations keeps the relevant information where it should be: with the test method. Instead of jumping back and forth from `TestUtil` and `AbstractModuleTestSupport` to check stack size and duration of the thread, we can easily verify the resources that we are working with while writing/ running the unit test.
2. We do not need to modify existing `verify...` methods. 
3. We do not need to create new `verify...` methods.
4. We actually capture, then re-throw any errors or exceptions, instead of counting on `FutureTask.get()` to show exceptions (a hack).
5. There is no need to worry about checking for return value of `verify` method (verify methods are void so this was also a hack to return `null` from the `Callable`).
6. We do not count on thread joining to throw exception (a hack that is not reliable).

Normally, I am not a big fan of lots of annotations on code, but I think that this case provides a lot of value and flexibility.

Could we just add `timeout` and `stackSize` to the `LimitedStackThread` interface so that we could do:

```java

@LimitedStackThread(timeout = 10000, stackSize = 147456)

```

@romani @rnveach @strkkk please share your thoughts.

Reply:
@rnveach

> I think we should use it everywhere

I do not see the value in limiting the stack size and duration for all unit tests. As noted at https://github.com/checkstyle/checkstyle/pull/10640#discussion_r690377269, the stack size will not be able to be the same for all tests. 

Additionally, if we end up using @pbludov 's approach mentioned above, this will require a great deal of refactoring in current unit tests.

I think it is better to keep the scope of this PR smaller, and restrict whatever solution that we end up with to the tests that require it.  If we want to limit the resources available to all unit tests, we should create a new issue for this.

Reply:
My branch is for the other issue mentioned. Not for this one. I am just using this PR's code as a base to see if issues will come up with it's usage.

The benefit to adding it to more tests is for us to know sooner if something will cause a stackoverflow before we get reports.

Reply:
> @LimitedStackThread(timeout = 10000, stackSize = 147456)

Perhaps a more general way is needed here. Something like

```java
@RunInIsolatedThread(name = "...", stackSize = 1,  priority = 1)
```